src/scvi/external/methylvi/_module.py

```diff
@@ -0,0 +1,269 @@
+    num_features_per_modality
+        Number of features corresponding to each modality
+    n_batch
+        Number of batches, if 0, no batch correction is performed.
+    n_hidden
+        Number of nodes per hidden layer
+    n_latent
+        Dimensionality of the latent space
+    n_layers
+        Number of hidden layers used for encoder and decoder NNs
+    dropout_rate
+        Dropout rate for neural networks
+    likelihood
+        One of
+        * ``'betabinomial'`` - BetaBinomial distribution
+        * ``'binomial'`` - Binomial distribution
+    dispersion
+        One of the following
+        * ``'region'`` - dispersion parameter of BetaBinomial is constant per region across cells
+        * ``'region-cell'`` - dispersion can differ for every region in every cell
+    """
+
+    def __init__(
+        self,
+        n_input: int,
+        modalities: Iterable[str],
+        num_features_per_modality: Iterable[int],
+        n_batch: int = 0,
+        n_hidden: int = 128,
+        n_latent: int = 10,
+        n_layers: int = 1,
+        dropout_rate: float = 0.1,
+        likelihood: Literal["betabinomial", "binomial"] = "betabinomial",
+        dispersion: Literal["region", "region-cell"] = "region",
+    ):
+        super().__init__()
+        self.n_latent = n_latent
+        self.n_batch = n_batch
+
+        self.latent_distribution = "normal"
+        self.dispersion = dispersion
+        self.likelihood = likelihood
+        self.modalities = modalities
+        self.num_features_per_modality = num_features_per_modality
+
+        cat_list = [n_batch]
+
+        self.z_encoder = Encoder(
+            n_input * 2,  # Methylated counts and coverage for each feature --> x2
+            n_latent,
+            n_cat_list=cat_list,
+            n_layers=n_layers,
+            n_hidden=n_hidden,
+            dropout_rate=dropout_rate,
+            return_dist=True,
+            var_activation=torch.nn.functional.softplus,  # Better numerical stability than exp
+        )
+
+        self.decoders = nn.ModuleDict()
+        for modality, num_features in zip(modalities, num_features_per_modality):
+            self.decoders[modality] = DecoderMETHYLVI(
+                n_latent,
+                num_features,
+                n_cat_list=cat_list,
+                n_layers=n_layers,
+                n_hidden=n_hidden,
+            )
+
+        if self.dispersion == "region":
+            self.px_gamma = torch.nn.ParameterDict(
+                {
+                    modality: nn.Parameter(torch.randn(num_features))
+                    for (modality, num_features) in zip(modalities, num_features_per_modality)
+                }
+            )
+
+    def _get_inference_input(self, tensors):
+        """Parse the dictionary to get appropriate args"""
+        mc = torch.cat(
+            [tensors[f"{x}_{METHYLVI_REGISTRY_KEYS.MC_KEY}"] for x in sorted(self.modalities)],
+            dim=1,
+        )
+        cov = torch.cat(
```

Comment:
I assumed they can have different nu,ber of features. Why does it work here?

Reply:
If they all have the same shape, and there is only one modality per cell. Shouldn't we ask the user to concatenate before inputing to methylVI. I assume those things are clear but I'm struggling with the meaning of modality.

Reply:
See my reply to your top-level comment for details here. Modalities aren't guaranteed to have the same same shape and we often will have more than one methylation "modality" per cell.

Reply:
Oh, you concatenate along the feature axis (not the cell axis). Makes sense now and thanks for the explaination.